xdsl/transforms/csl_wrapper_to_csl.py

```diff
@@ -0,0 +1,144 @@
+from dataclasses import dataclass
+
+from xdsl.builder import ImplicitBuilder
+from xdsl.context import MLContext
+from xdsl.dialects import arith, builtin, scf
+from xdsl.dialects.csl import csl, csl_wrapper
+from xdsl.ir import Block, BlockArgument, Operation, Region, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class RemoveUnusedYield(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_wrapper.YieldOp, rewriter: PatternRewriter, /):
+        parent = op.parent_op()
+        assert isa(parent, csl_wrapper.ModuleOp)
+        if op.parent_region() == parent.program_module:
+            rewriter.erase_matched_op()
+
+
+@dataclass(frozen=True)
+class ExtractLayoutProgramModule(RewritePattern):
+    prog_name: str
+
+    @staticmethod
+    def replace_arg(
+        rewriter: PatternRewriter, arg: BlockArgument, param: Operation | SSAValue
+    ):
+        for use in arg.uses:
+            rewriter.handle_operation_modification(use.operation)
+        arg.replace_by(SSAValue.get(param))
+        rewriter.erase_block_argument(arg)
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_wrapper.ModuleOp, rewriter: PatternRewriter, /):
+        module_block = Block()
+
+        outer_loop_block = Block()
+        outer_loop_block.insert_arg(builtin.IntegerType(32), 0)
+        x = outer_loop_block.args[0]
+
+        inner_loop_block = Block()
+        inner_loop_block.insert_arg(builtin.IntegerType(32), 0)
+        y = inner_loop_block.args[0]
+
+        new_args = list[SSAValue]()
+
+        assert isa(yield_op := op.layout_module.block.last_op, csl_wrapper.YieldOp)
+        yield_op.detach()
+
+        with ImplicitBuilder(module_block):
+            const_0 = arith.Constant.from_int_and_width(0, builtin.IndexType())
+            const_1 = arith.Constant.from_int_and_width(1, builtin.IndexType())
+
+            const_width = arith.Constant(op.width)
+            param_width = csl.ParamOp("width", op.width.type, const_width)
```

Comment:
Having separate `const_width` and `param_width` values means that the values can diverge. For instance, csl.param values have a default but can be passed a different value from command line, which would not affect the value in const_width. Best to remove const_width and only use param_width, same for *_height

Reply:
Ok actually the confusion might have been the use of const_width in the loop.

Reply:
Fixed